### PR TITLE
Implement hashing to curve

### PR DESCRIPTION
**Pull request by onvej-sl**
This pull requests implements hashing to a curve that is compatible with the suite [`P256_XMD:SHA-256_SSWU_RO_`](https://www.rfc-editor.org/rfc/rfc9380.html#name-suites-for-nist-p-256) from RFC9380.

### CODE REVIEW

**Comment by onvej-sl** [comment]
@andrewkozlik is going to do a brief overall review.

@matejcik, @hiviah, I need at least one of you to check the memory safety (particularly the out-of-bound reads and writes) of these functions:
  * `expand_message_xmd_sha256`
  * `hash_to_field`
  * `hash_to_curve`

File: crypto/hash_to_curve.c [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+  }
+
+  return true;
+}
+
+bool hash_to_curve_optiga(const uint8_t input[32], uint8_t public_key[65]) {
+  char dst[] = "OPTIGA-SECRET-V0-P256_XMD:SHA-256_SSWU_RO_";
+  curve_point point = {0};
+
+  if (!hash_to_curve_p256(input, 32, (uint8_t *)dst, sizeof(dst), &point)) {
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply crypto/hash_to_curve.c]: Definitely not intentional. Fixed in https://github.com/trezor/trezor-firmware/pull/3293/commits/9fcf672176e8ad694278900eeef6986bbdeac653.

File: crypto/hash_to_curve.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef __HASH_TO_CURVE_H__
+#define __HASH_TO_CURVE_H__
+
+#include "bignum.h"
```
Comment (andrewkozlik) [inline crypto/hash_to_curve.h:24]: I think we don't need to include bignum.h.

**Review by matejcik — COMMENTED** [review]
C-wise LGTM.

File: crypto/hash_to_curve.c [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+// https://www.rfc-editor.org/rfc/rfc9380.html#name-hash_to_field-implementatio
+static bool hash_to_field(const uint8_t *msg, size_t msg_len,
+                          const uint8_t *dst,  // domain separation tag
+                          const size_t dst_len, size_t expansion_len,
+                          const bignum256 *prime,
+                          bool expand(const uint8_t *, size_t, const uint8_t *,
+                                      size_t, uint8_t *, size_t),
+                          bignum256 *out, size_t out_len) {
+  const size_t max_expansion_len = 64;
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply crypto/hash_to_curve.c:40]: * `#define` has global "scope".
* I expect the compiler to inline the variable.

**Pull request merged**